executor/write.go

```diff
@@ -315,6 +299,148 @@ func (e *DeleteExec) deleteSingleTable(goCtx goctx.Context) error {
+	batchDelete := e.ctx.GetSessionVars().BatchDelete && !e.ctx.GetSessionVars().InTxn()
+	batchDMLSize := e.ctx.GetSessionVars().DMLBatchSize
+	fields := e.children[0].Schema().GetTypes()
+	for {
+		chk := e.children[0].newChunk()
+		err := e.children[0].NextChunk(goCtx, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		for chunkRow := chk.Begin(); chunkRow != chk.End(); chunkRow = chunkRow.Next() {
+			if batchDelete && rowCount >= batchDMLSize {
+				if err = e.ctx.NewTxn(); err != nil {
+					// We should return a special error for batch insert.
+					return ErrBatchInsertFail.Gen("BatchDelete failed with error: %v", err)
+				}
+				rowCount = 0
+			}
+
+			datumRow := chunkRow.GetDatumRow(fields)
+			err = e.deleteOneRow(tbl, handleCol, datumRow)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			rowCount++
+		}
+	}
+
+	return nil
+}
+
+func (e *DeleteExec) initialMultiTableTblMap() {
+	e.tblMap = make(map[int64][]*ast.TableName, len(e.Tables))
+	for _, t := range e.Tables {
+		e.tblMap[t.TableInfo.ID] = append(e.tblMap[t.TableInfo.ID], t)
+	}
+}
+
+func (e *DeleteExec) getColPosInfos(schema *expression.Schema) []tblColPosInfo {
+	var colPosInfos []tblColPosInfo
+	// Extract the columns' position information of this table in the delete's schema, together with the table id
+	// and its handle's position in the schema.
+	for id, cols := range schema.TblID2Handle {
+		tbl := e.tblID2Table[id]
+		for _, col := range cols {
+			if !e.matchingDeletingTable(id, col) {
+				continue
+			}
+			offset := getTableOffset(schema, col)
+			end := offset + len(tbl.Cols())
+			colPosInfos = append(colPosInfos, tblColPosInfo{tblID: id, colBeginIndex: offset, colEndIndex: end, handleIndex: col.Index})
+		}
+	}
+	return colPosInfos
+}
+
+func (e *DeleteExec) composeTblRowMap(tblRowMap tableRowMapType, colPosInfos []tblColPosInfo, joinedRow Row) {
+	// iterate all the joined tables, and got the copresonding rows in joinedRow.
+	for _, info := range colPosInfos {
+		if tblRowMap[info.tblID] == nil {
+			tblRowMap[info.tblID] = make(map[int64][]types.Datum)
+		}
+		handle := joinedRow[info.handleIndex].GetInt64()
+		// tblRowMap[info.tblID][handle] hold the row datas binding to this table and this handle.
+		tblRowMap[info.tblID][handle] = joinedRow[info.colBeginIndex:info.colEndIndex]
+	}
+}
+
+func (e *DeleteExec) deleteMultiTablesByChunk(goCtx goctx.Context) error {
+	if len(e.Tables) == 0 {
+		return nil
+	}
+
+	e.initialMultiTableTblMap()
+	colPosInfos := e.getColPosInfos(e.children[0].Schema())
+	tblRowMap := make(tableRowMapType)
+	fields := e.children[0].Schema().GetTypes()
+	for {
+		chk := e.children[0].newChunk()
+		err := e.children[0].NextChunk(goCtx, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chk.NumRows() == 0 {
+			break
+		}
+
+		for joinedChunkRow := chk.Begin(); joinedChunkRow != chk.End(); joinedChunkRow = joinedChunkRow.Next() {
```

Comment:
what does `joinedChunkRow` mean ?

Reply:
delete multi tables actually delete the data from joined tables.